### PR TITLE
163 crud operations on purchase module

### PR DESCRIPTION
**Pull request by pavanpulicherla18345**

### CODE REVIEW

**Review by QodeFarm — COMMENTED** [review]
Review thw comments given to Pramod and Bhagyaraj and address them here.

File: apps/purchase/views.py [hunk 1]
```diff
@@ -1,98 +1,266 @@
...
+                "purchase_order": purchaseorder_data,
+                "purchase_order_items": purchaseorder_items_data,
+                "order_attachments": orderattachments_data,
+                "order_shipments": ordershipments_data
+            }
+            return Response({
+                'count': 1,
+                'message': 'Record updated successfully',
+                'data': custom_data
+            },status=status.HTTP_200_OK)
```
Comment (QodeFarm) [inline apps/purchase/views.py:243]: Use build_respopnse function here.

File: apps/purchase/views.py [hunk 2]
```diff
@@ -1,98 +1,266 @@
...
+        else:
+            return Response({
+                'count': 0,
+                'message': 'Record updation failed',
+                'errors': serializer.errors
+            }, status=status.HTTP_400_BAD_REQUEST)
```
Comment (QodeFarm) [inline apps/purchase/views.py:249]: Use build_respopnse function here.

File: apps/purchase/views.py [hunk 3]
```diff
@@ -1,98 +1,266 @@
...
+##---------- ONE API - MULTIPLE API CALLS (CRUD OPERATIONS) ---------------
+
+class purchaseOrdersCreateView(generics.GenericAPIView, mi.ListModelMixin, mi.CreateModelMixin, mi.RetrieveModelMixin, mi.UpdateModelMixin, mi.DestroyModelMixin):
+    queryset = PurchaseOrders.objects.all()
+    serializer_class = PurchaseOrdersSerializer
+ 
+    def get(self, request, *args, **kwargs):
+        if 'pk' in kwargs:
+            return self.retrieve(request, *args, **kwargs)  # Retrieve a single instance
+        return list_all_objects(self, request, *args, **kwargs)
```
Comment (QodeFarm) [inline apps/purchase/views.py:117]: Use single return statement.

File: apps/purchase/views.py (hunk 2 above)
Comment (pavanpulicherla18345) [reply apps/purchase/views.py:249]: Okay Thulasi.

File: apps/purchase/views.py (hunk 1 above)
Comment (pavanpulicherla18345) [reply apps/purchase/views.py:243]: Okay Thulasi.

File: apps/purchase/views.py [hunk 4]
```diff
@@ -1,98 +1,428 @@
...
+                "order_shipments": shipments_data
+            }
+            response_data = build_response(1, "Record updated successfully", custom_data, status.HTTP_200_OK)
+        else:
+            logger.error("Error in PurchaseOrderViewSet")
+            response_data = build_response(0, "Record updation failed", [serializer.errors], status.HTTP_400_BAD_REQUEST)
+        
+        return response_data
+    
+    def put(self, request, pk, *args, **kwargs):
```
Comment (QodeFarm) [inline apps/purchase/views.py:382]: why do you have 2 put methods in this file?

File: apps/purchase/models.py [hunk 5]
```diff
@@ -89,8 +89,8 @@ class PurchaseInvoiceOrders(OrderNumberMixin):
     vendor_agent_id = models.ForeignKey(VendorAgent, on_delete=models.CASCADE, null=True, default=None, db_column = 'vendor_agent_id')
```
Comment (QodeFarm) [inline apps/purchase/models.py:90]: Can you add below 2 columns  to DB, Model and code as well,

shipping_address VARCHAR(1024),
billing_address VARCHAR(1024),

These 2 fields should be part of payload(CREATE and PUT)

Comment (pavanpulicherla18345) [reply apps/purchase/models.py:90]: Okay Thulasi.

**Pull request merged**